crates/sui-core/src/authority/consensus_quarantine.rs

```diff
@@ -330,3 +334,647 @@ impl ConsensusCommitOutput {
+    pub(super) deferred_transactions:
+        Mutex<BTreeMap<DeferralKey, Vec<VerifiedSequencedConsensusTransaction>>>,
+    // user_signatures_for_checkpoints is written to by consensus handler and read from by checkpoint builder
+    // The critical sections are small in both cases so a DashMap is probably not helpful.
+    pub(super) user_signatures_for_checkpoints:
+        Mutex<HashMap<TransactionDigest, Vec<GenericSignature>>>,
+}
+
+impl ConsensusEphemeralOutput {
+    pub(crate) fn new(
+        epoch_start_configuration: &EpochStartConfiguration,
+        tables: &AuthorityEpochTables,
+    ) -> Self {
+        let shared_version_assignments =
+            Self::get_all_shared_version_assignments(epoch_start_configuration, tables);
+
+        let deferred_transactions = tables
+            .get_all_deferred_transactions()
+            .expect("load deferred transactions cannot fail");
+
+        let user_signatures_for_checkpoints = tables
+            .get_all_user_signatures_for_checkpoints()
+            .expect("load user signatures for checkpoints cannot fail");
+
+        Self {
+            shared_version_assignments: shared_version_assignments.into_iter().collect(),
+            deferred_transactions: Mutex::new(deferred_transactions),
+            user_signatures_for_checkpoints: Mutex::new(user_signatures_for_checkpoints),
+        }
+    }
+
+    // Used to read pre-existing shared object versions from the database after a crash.
+    // TODO: remove this once all nodes have upgraded to data-quarantining.
+    fn get_all_shared_version_assignments(
+        epoch_start_configuration: &EpochStartConfiguration,
+        tables: &AuthorityEpochTables,
+    ) -> Vec<(
+        TransactionKey,
+        Vec<(ConsensusObjectSequenceKey, SequenceNumber)>,
+    )> {
+        if epoch_start_configuration.use_version_assignment_tables_v3() {
+            tables
+                .assigned_shared_object_versions_v3
+                .safe_iter()
+                .collect::<Result<_, _>>()
+                .expect("db error")
+        } else {
+            tables
+                .assigned_shared_object_versions_v2
+                .safe_iter()
+                .collect::<Result<Vec<_>, _>>()
+                .expect("db error")
+                .into_iter()
+                .map(|(key, value)| {
+                    (
+                        key,
+                        value
+                            .into_iter()
+                            .map(|(id, v)| ((id, SequenceNumber::UNKNOWN), v))
+                            .collect::<Vec<_>>(),
+                    )
+                })
+                .collect()
+        }
+    }
+}
+
+/// ConsensusOutputQuarantine holds outputs of consensus processing in memory until the checkpoints
+/// for the commit have been certified.
+pub(crate) struct ConsensusOutputQuarantine {
+    // Output from consensus handler
+    output_queue: VecDeque<ConsensusCommitOutput>,
```

Comment:
Since this seemingly can grow unbounded, should we have a metric to track how many items are there?
(And similarly perhaps for shared_object_next_versions?)

Reply:
good idea - it should be bounded due to consensus back pressure, although its somewhat indirect, so i agree monitoring this is a good idea